R/deploymentTarget.R

```diff
@@ -1,210 +1,66 @@
 # calculate the deployment target based on the passed parameters and
 # any saved deployments that we have
-deploymentTarget <- function(appPath,
+deploymentTarget <- function(appPath = ".",
                              appName = NULL,
                              appTitle = NULL,
                              appId = NULL,
                              account = NULL,
-                             server = NULL) {
-  # read existing accounts
-  accounts <- accounts(server)[, "name"]
-  if (length(accounts) == 0) {
-    stopWithNoAccount()
-  }
+                             server = NULL,
+                             error_call = caller_env()) {
+
+  appDeployments <- deployments(
+    appPath = appPath,
+    nameFilter = appName,
+    accountFilter = account,
+    serverFilter = server
+  )
 
-  # validate account if provided
-  if (!is.null(account)) {
-    if (!account %in% accounts) {
-      stop(
-        paste("Unknown account name '", account, "' (you can use the ",
-          "setAccountInfo function to add a new account)",
-          sep = ""
-        ),
-        call. = FALSE
-      )
+  if (nrow(appDeployments) == 0) {
+    fullAccount <- findAccount(account, server)
+    if (is.null(appName)) {
+      appName <- generateAppName(appTitle, appPath, account, unique = FALSE)
     }
-  }
-
-  # read existing deployments
-  appDeployments <- deployments(appPath = appPath)
-
-
-  # if appTitle specified but not appName, generate name from title
-  if (is.null(appName) && !is.null(appTitle) && nzchar(appTitle)) {
```

Comment:
This is now only done on first deploy as documented. And I'm pretty sure the previous behaviour was incorrect as it meant that setting a new `appTitle` would generally result in publishing a new app, rather than updating an existing app.

Reply:
Could you check what values are provided by the IDE? Its conventions have changed over time. In particular, what happens when going through the "Other Destination" workflow? Is that different from a "republish" workflow?

I don't recall what of name/title is given by the IDE. It would be good to capture that expectation.

We'll probably want to retest a handful of "new deployment" versus "overwrite deployment" workflows. CC @ChaitaC

Reply:
Here's what I see in the "Other destination" dialog — it like the UI allows you to choose the `appTitle`, which appears to be auto generated from the doc name:

<img width="598" alt="Screenshot 2023-02-13 at 16 34 47" src="https://user-images.githubusercontent.com/4196/218589913-076238ec-e86b-48ba-a8cf-85a02c40c191.png">

It's not clear what it users for the `appName`. I'll need to do some more exploration.

The republish dialog only allows me to change which files are published.

Reply:
Note that the id also uses the output of `applications()` to find the `appId` based on the `appName`, so that even if `deployments()` is empty, it'll still attempt to update an existing bundle.

Reply:
I've made an issue https://github.com/rstudio/rsconnect/issues/669 to track logging more information so it's easier to verify this. I'll do that in a separate PR because this one is already big.